app/models/concerns/impressionable.rb

```diff
@@ -1,79 +1,115 @@
 module Impressionable
   extend ActiveSupport::Concern
 
-  def total_impressions_count(start_date = nil, end_date = nil)
-    return daily_impressions_counts(start_date, end_date).sum if start_date && end_date
-    key = "#{cache_key}/#{__method__}/#{Date.current.cache_key minutes_cached: 15}"
-    Rails.cache.fetch(key) { impressions.count }.to_i
-  end
-
-  def total_impressions_per_mille
-    total_impressions_count.to_i / 1_000.to_f
+  def daily_impressions_counts(start_date = nil, end_date = nil, scoped_by: nil)
+    start_date = Date.coerce(start_date)
+    end_date = Date.coerce(end_date || start_date)
+    key = "#{cache_key}/#{__method__}/#{start_date.cache_key}-#{end_date.cache_key}/#{scoped_by&.cache_key}"
+    Rails.cache.fetch key do
+      counts_by_date = daily_summaries.between(start_date, end_date).scoped_by(scoped_by)
+        .pluck(:displayed_at_date, :impressions_count)
+        .each_with_object({}) { |row, memo| memo[row[0]] = row[1] }
+      dates = (start_date..end_date).to_a
+      missing_dates = dates - counts_by_date.keys
+      if missing_dates.present?
+        CreateDailySummariesJob.perform_later self, start_date.iso8601, end_date.iso8601, scoped_by
+        impressions.on(*missing_dates).scoped_by(scoped_by).group(:displayed_at_date).count
+          .each { |date, count| counts_by_date[date] ||= count }
+      end
+      dates.map { |date| counts_by_date[date] ||= 0 }
```

Comment:
Why are we excluding 0's from this response? I'm just curious.

Reply:
Not sure I follow you on this question.